### PR TITLE
Improve password policy

### LINKED ISSUE #11752: Weak password policy
Thanks Jhon Ken

Some websites prevents the users to use password constructed of character combinations that otherwise meet company policy, but should no longer be used because they have been deemed insecure for one or more reasons, such as being easily guessed, following a common pattern, or public disclosure from previous data breaches. 

Steps to reproduce: 
- Go to signup page
- Create an account by typing your email address and your password same as your email.
- Hit "Sign Up for " button.
- Your account will be created.

Some policies suggest or impose requirements on what type of password a user can choose, such as:

A key concern when using passwords for authentication is password strength. A "strong" password policy makes it difficult or even improbable for one to guess the password through either manual or automated means. The following characteristics define a strong password:

### Implement Proper Password Strength Controls

A key concern when using passwords for authentication is password strength. A "strong" password policy makes it difficult or even improbable for one to guess the password through either manual or automated means. The following characteristics define a strong password:

- Password Length
    - **Minimum** length of the passwords should be **enforced** by the application. Passwords **shorter than 8 characters** are considered to be weak ([NIST SP800-63B](https://pages.nist.gov/800-63-3/sp800-63b.html)).
    - **Maximum** password length should not be set **too low**, as it will prevent users from creating passphrases. A common maximum length is 64 characters due to limitations in certain hashing algorithms, as discussed in the [Password Storage Cheat Sheet](Password_Storage_Cheat_Sheet.md#maximum-password-lengths). It is important to set a maximum password length to prevent [long password Denial of Service attacks](https://www.acunetix.com/vulnerabilities/web/long-password-denial-of-service/).
- Do not silently truncate passwords. The [Password Storage Cheat Sheet](Password_Storage_Cheat_Sheet.md#maximum-password-lengths) provides further guidance on how to handle passwords that are longer than the maximum length.
- Allow usage of **all** characters including unicode and whitespace. There should be no password composition rules limiting the type of characters permitted.
- Ensure credential rotation when a password leak, or at the time of compromise identification.
- Include password strength meter to help users create a more complex password and block common and previously breached passwords
    - [zxcvbn library](https://github.com/dropbox/zxcvbn) can be used for this purpose. (Note that this library is no longer maintained)
    - [Pwned Passwords](https://haveibeenpwned.com/Passwords) is a service where passwords can be checked against previously breached passwords. You can host it yourself or use [API](https://haveibeenpwned.com/API/v3#PwnedPasswords).
Refer to this link: https://en.wikipedia.org/wiki/Password_policy

Follow OWASP cheatsheets: https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Authentication_Cheat_Sheet.md#password-complexity

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Update password policy<br>- At least 8 characters<br>- 1 uppercase letter<br>- 1 lowercase letter<br>- 1 digit<br>- 1 special chars<br>Old password are still working<br>Also update Passy to latest version (even if this one is abandoned)
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #11752 
| How to test?  | You should be able to install / create customer account, create employee, but if you have an old installation passwords are still valid.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12824)

### CODE REVIEW

File: classes/Validate.php [hunk 1]
```diff
@@ -523,7 +523,7 @@ public static function isHashedPassword($hashedPasswd)
 
     public static function isPasswdAdmin($passwd)
     {
-        return Validate::isPlaintextPassword($passwd, Validate::ADMIN_PASSWORD_LENGTH);
+        return self::isPlaintextPassword($passwd, Validate::ADMIN_PASSWORD_LENGTH);
```
Comment (eternoendless) [inline classes/Validate.php]:
```suggestion
        return static::isPlaintextPassword($passwd, Validate::ADMIN_PASSWORD_LENGTH);
```

File: classes/Validate.php [hunk 2]
```diff
@@ -500,7 +500,7 @@ public static function isPasswd($passwd, $size = Validate::PASSWORD_LENGTH)
     public static function isPlaintextPassword($plaintextPasswd, $size = Validate::PASSWORD_LENGTH)
     {
         // The password lenght is limited by `password_hash()`
-        return Tools::strlen($plaintextPasswd) >= $size && Tools::strlen($plaintextPasswd) <= 72;
+        return preg_match('/^(?=.*[A-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^\w])\S{' . $size . ',72}$/', $plaintextPasswd);
```
Comment (eternoendless) [inline classes/Validate.php]: This regex enforces the policy I suppose? Could you please add a comment like "Ensure the password has at least one of x, one of y and one of z"?

Also, I'm wondering why did you use A-z? If you look at an ASCII table, you'll notice that A-z includes the following characters as well:

	[ \ ] ^ _ `

Also, any non A-z character like `é` or `ç` will count as a non word character.

File: classes/Validate.php [hunk 3]
```diff
@@ -26,7 +26,7 @@
 class ValidateCore
 {
     const ADMIN_PASSWORD_LENGTH = 8;
-    const PASSWORD_LENGTH = 5;
+    const PASSWORD_LENGTH = 8;
```
Comment (eternoendless) [inline classes/Validate.php]: According to OWASP, 10 is a better minimum

File: classes/form/CustomerForm.php [hunk 4]
```diff
@@ -127,6 +127,19 @@ public function validate()
             $birthdayField->setValue($dateBuilt->format('Y-m-d'));
         }
 
+        // password check
+        $passwordValue = $this->getValue('password');
+        if (!empty($passwordValue) && !Validate::isPasswd($passwordValue)) {
+            $passwordField = $this->getField('password');
+            $passwordField->addError(
+                $this->translator->trans(
+                    'At least 8 characters long, 1 lowercase letter, 1 uppercase letter, 1 digit and 1 special char',
```
Comment (eternoendless) [inline classes/form/CustomerForm.php]:
```suggestion
                    'At least %number% characters long, 1 lowercase letter, 1 uppercase letter, 1 digit and 1 symbol (eg. @!?^&)',
```

Shouldn't this message start with something like: "Your password should be at least... and contain at least..."?

File: install-dev/controllers/http/configure.php [hunk 5]
```diff
@@ -109,7 +109,11 @@ public function validate()
         // Check passwords
         if ($this->session->admin_password) {
             if (!Validate::isPasswdAdmin($this->session->admin_password)) {
-                $this->errors['admin_password'] = $this->translator->trans('The password is incorrect (must be alphanumeric string with at least 8 characters)', array(), 'Install');
+                $this->errors['admin_password'] = $this->translator->trans(
+                    'The password is incorrect (must be alphanumeric string with at least 8 characters, 1 uppercase letter, 1 lowercase letter, 1 digit and 1 special char)',
```
Comment (eternoendless) [inline install-dev/controllers/http/configure.php]: Same comment as in the other wording

File: themes/classic/templates/_partials/form-fields.tpl [hunk 6]
```diff
@@ -136,10 +136,10 @@
             <input
               class="form-control js-child-focus js-visible-password"
               name="{$field.name}"
-              title="{l s='At least 5 characters long' d='Shop.Forms.Help'}"
+              title="{l s='At least 8 characters long, 1 lowercase letter, 1 uppercase letter, 1 digit and 1 special char' d='Shop.Forms.Help'}"
```
Comment (eternoendless) [inline themes/classic/templates/_partials/form-fields.tpl]: Again, the number of characters should be injected.
```suggestion
              title="{l s='At least 8 characters long, 1 lowercase letter, 1 uppercase letter, 1 digit and 1 symbol (eg. @!?^&)' d='Shop.Forms.Help'}"
```

File: themes/classic/templates/_partials/form-fields.tpl [hunk 7]
```diff
@@ -136,10 +136,10 @@
...
               type="password"
               value=""
-              pattern=".{literal}{{/literal}5,{literal}}{/literal}"
+              pattern="{literal}((?=.*[A-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^\w])\S{8,72}|.{5,72}){/literal}"
```
Comment (eternoendless) [inline themes/classic/templates/_partials/form-fields.tpl]: Same comment about A-z

**Review by eternoendless — CHANGES_REQUESTED** [review]

**Comment by MathiasReker** [comment]
How about adding a button next to the password field that can generate a valid password?

**Comment by PierreRambaud** [comment]
@MathiasReker Already exists in back office. Could be cool for front office but in a new PR :)

File: classes/Validate.php (hunk 2 above)
Comment (PierreRambaud) [reply classes/Validate.php]: For `A-z` it's a mistake I wanted to write `a-z`.
For `é` and `ç`, I consider them as special characters with `(?=.*[^a-zA-Z0-9])`

File: themes/classic/templates/_partials/form-fields.tpl (hunk 7 above)
Comment (PierreRambaud) [reply themes/classic/templates/_partials/form-fields.tpl]: I inject with `Validate::PASSWORD_LENGTH` this form is too generic to inject a variable with smarty :/

**Comment by rdy4ever** [comment]
Hi @PierreRambaud
I know that security wise this makes sense, but I also know that this will bother the hell out of most customers, increasing cart abandonment quite a fair bit. 
May I please suggest to make this optional, with an admin toggle?
By default, it could be like this, but please (pretty please) add in settings/customers a toogle for "strong customer password policy".
Also, how secure must a customer account be, after all? We don't store any credit card info in the customer account, so there really isn't much interest to "hack" a customer account, wright?
By all means, this password policy is good for admin accounts, but for customers this will be very annoying. And we don't want to annoy our customers, am I wright?

**Comment by PierreRambaud** [comment]
We value the security of your customers and think this is very important, PrestaShop would be responsible.
Not sure this will bother customers, for a customer it's always better to see a website that take in consideration security than allowing '11111' as password :/

But of course we can plan to add an option to `Activate weak password policy`, depends of what @colinegin @eternoendless and @PrestaShop/prestashop-core-developers say

**Comment by matks** [comment]
> Also, how secure must a customer account be, after all? We don't store any credit card info in the customer account, so there really isn't much interest to "hack" a customer account, wright?

@rdy4ever First name last name, phone number and adresses are critical data too 😉 they can be used to impersonate someone and take control of some of his accounts (the scheme is to use these data to answer the "I forgot my password - answer a question" mecanism).

PrestaShop has to protect both customers and admin users, but this is an opensource project: you can always modify the behavior if needed 😉 . I can think of 3 solutions to provide a practical system to easily allow you to use a less restrictive password policy:
- keep old password policy (but disable it and enable the strong policy by default) and allow it to be enabled from the BO
- provide an Interface PasswordPolicyManagerInterface for the class responsible for the password policy so that you can replace it with your own implementation
- add a hook able to authorize/unanthorize a password so you can use it to modify the shop password policy through a module (actually, this could be used to enable SSO for front-office ?)

**Comment by rblaurin** [comment]
Do this PR cover #12340? Maybe it could be closed?

**Comment by rblaurin** [comment]
> I can think of 3 solutions to provide a practical system to easily allow you to use a less restrictive password policy:

I like the solutions. Would It be excessive to use something like this? https://github.com/rollerworks/PasswordStrengthValidator

**Comment by PierreRambaud** [comment]
@rblaurin Not sure we can use this because of legacy parts :/

**Comment by rdy4ever** [comment]
OK, I know everyone might hate me for this, but, why must a prestashop client account impose a stronger password policy than Google? Why "at least 8 characters" is good enough for Google (that also uses that account in sensitive places like Google Ads where you do have credit card info saved and someone could just promote their site with your money - not to mention that you can basically reset other passwords if you have access to an email account) but a simple e-shop client account must ask for: 

>- 1 uppercase letter
>- 1 lowercase letter
>- 1 digit
>- 1 special chars

I mean, I am not against this whole stronger password policy, but make it optional for the merchant. If you add this in 1.7.6 without an opt-out in admin, I can guarantee a whole bunch of merchants will stop upgrading their shops, making them even less secure.

image

**Comment by PierreRambaud** [comment]
> why must a prestashop client account impose a stronger password policy than Google

Because we don't have two factor authentication, no shadow ban, no captcha in case of spam, etc.
They improve their security in many other ways (Saw this with them few months ago).
The best practice is following the OWASP recommendations.

**Comment by matks** [comment]
> OK, I know everyone might hate me for this

Eyh, dont worry github is an area open for discussion 😛you can disagree with us without us hating you, it's OK to have different opinions 😉

**Comment by rblaurin** [comment]
> Eyh, dont worry github is an area open for discussion 😛you can disagree with us without us hating you

Talk for yourself, I hate him! 😛
Just kidding. Actually I agree with him when he says it should be optional (but default). At least for front office.

**Comment by matks** [comment]
@PierreRambaud I just came back to this discussion from https://github.com/PrestaShop/docker/pull/157. I guess, as it went sale, that we can close this PR for now ?

**Comment by PierreRambaud** [comment]
I keep thinking it's a security breach. We shouldn't let customers / users to set a password like `1234`.
And because we have to do our best to follow best practices not only with the code. Security is maybe more important.

**Comment by matks** [comment]
> I keep thinking it's a security breach. We shouldn't let customers / users to set a password like `1234`.
> And because we have to do our best to follow best practices not only with the code. Security is maybe more important.

Then ... let's update and merge this PR for 1.7.7 😉

We can use one of the suggestions in https://github.com/PrestaShop/PrestaShop/pull/12824#issuecomment-470866155 to make it easier to transition for people

**Comment by colinegin** [comment]
I'm sorry there was no clear decision on this important topic. 
Ensure security is key, both for our merchants, and their customers, and I think your PR is a really good opportunity to improve it, as long as it does not impact current passwords. 
And as @rdy4ever and @rblaurin suggested, it's a good idea to add this as an optional setting. I would recommend to put it in Shop parameters > Customer settings as there is already an option for passwords. Wdyt @PierreRambaud @matks ?

**Comment by PierreRambaud** [comment]
@colinegin Current passwords are still working with this PR.
Making this kind of verification optional isn't a good idea. It will duplicate code :/ maybe we want to offer the capability to configure passwords rules?

**Comment by colinegin** [comment]
@PierreRambaud yes that could be interesting, but how could we manage this ?

**Comment by PierreRambaud** [comment]
A possibility could be to have a checkbox with what characters we want in the password:
- at least 1 digit
- at least 1 lowercase
- at least 1 uppercase
- at least 1 special chars

and of course

- minimum length
- maximum length

**Comment by matks** [comment]
Heard at ForumPHP 2019 from @Paulmolin ([this guy](https://event.afup.org/forumphp2019-interview-paul-molin/)) who is a member of the OWASP:

"In 2019, the OWASP does not suggest anymore forcing users to use a uppercase letter, a digit or a special char ; because we have observed that this makes passwords harder to remember, and consequently users start using the same password everywhere which is a bad security practice.

Instead, we suggest forcing passwords long enough, and suggesting to users to use a password manager."

@Paulmolin I hope I did not quote you badly 😅feel free to tell me if I did

File: classes/Validate.php [hunk 8]
```diff
@@ -483,7 +483,12 @@ public static function isReference($reference)
      */
     public static function isPasswd($passwd, $size = Validate::PASSWORD_LENGTH)
     {
-        return self::isPlaintextPassword($passwd, $size);
+        // Password is already a hashed no need to check
+        if (static::isHashedPassword($passwd)) {
+            return true;
+        }
+
+        return static::isPlaintextPassword($passwd, $size);
```
Comment (Progi1984) [inline classes/Validate.php:529]: @PierreRambaud Can't you extract this two methods and move it to a service ?

Comment (PierreRambaud) [reply classes/Validate.php:529]: It's related to Legacy context, and services are not available in the FO :thinking:

**Comment by ksaandev** [comment]
Hi,

I think it interesting to see OWASP suggestions 😉 
https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls

**Comment by PierreRambaud** [comment]
> Hi,
> 
> I think it interesting to see OWASP suggestions 
> https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls

This is what we are planning to do. OWASP changes its documentation since last year.  :sweat_smile:

**Comment by PierreRambaud** [comment]
Still waiting for the new Security page to integrate this new behavior

**Comment by PierreRambaud** [comment]
The new security page will not be merged, I'm closing it as this one will also not be merged :sweat_smile:

**Comment by PierreRambaud** [comment]
@eternoendless this PR is now useless, a lot of things have changed, and it should be configurable in the BO through the Security page.

**Pull request closed without merging**